### PR TITLE
Feature/hungarian to utils

### PR DESCRIPTION
**Pull request by mrlukasbos**
This PR moves the hungarian algorithm to utils, since that is a much better place for it. A large part of the code in this PR is from a library actually, which was previously in roboteam_ai. Although I made some extensions to this library so it better fits our needs. These where in roboteam_ai as well but I made them a little bit more generic now. besides that it is now fully tested.

I hope the code is clear enough that it has a nice interface for everyone to use it. suggestions are welcome.

Changelog:
- Moved code over to utils
- All functions are now static (all functions are pure!), making the hungarian easier to call and test
- some variable renaming.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -51,12 +53,11 @@ target_include_directories(roboteam_utils
 
 target_link_libraries(roboteam_utils
         PUBLIC roboteam_proto
-        #gtest
         )
 target_compile_options(roboteam_utils PRIVATE
         --std=c++17
         -Wall
-        -Werror
+#        -Werror
```
Comment (rolfvdhulst) [inline CMakeLists.txt]: The point is we don't want ugly code in this repo. If you need to comment this to make sure it compiles ....

Comment (rolfvdhulst) [reply CMakeLists.txt]: So I just compiled locally with the flag and it did not give an error (which is good). I don't see why we should remove it here.

Comment (rolfvdhulst) [reply CMakeLists.txt]: So, just change this line back and you should be fine.

Comment (mrlukasbos) [reply CMakeLists.txt]: haha, I still have to look into this. Had warnings locally.

Comment (mrlukasbos) [reply CMakeLists.txt]: will fix

Comment (mrlukasbos) [reply CMakeLists.txt]: already found the issue. The protobuf library does not link for me without warnings. It seems a cannot do something about it. https://github.com/protocolbuffers/protobuf/issues/3128

Comment (mrlukasbos) [reply CMakeLists.txt]: With all due respect I now don't want to change it; with this flag I can't run anything anymore

Comment (rolfvdhulst) [reply CMakeLists.txt]: Ah. It's ok for now I guess but let's track that issue and enable it again once it's fixed. That's indeed very silly.

**Pull request merged**